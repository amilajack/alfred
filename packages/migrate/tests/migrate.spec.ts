import path from 'path';
import espree from 'espree';
import { expect as chaiExpect } from 'chai';
import Providers, { handleInput } from '../src';

jest.setTimeout(20000);

function parseCodeSnippets(codeSnippets: Array<string> | void): void {
  if (codeSnippets) {
    codeSnippets.map(code =>
      espree.parse(code, {
        sourceType: 'module',
        ecmaVersion: 2018
      })
    );
  }
}

describe('Migrate', () => {
  const migrateOpts = {
    packageJsonPath: require.resolve(
      './fixtures/express-hello-world/package.json'
    ),
    root: path.join(__dirname, './fixtures/express-hello-world'),
    write: false,
    unsafe: false,
    verbose: false
  };

  it('should migrate basic file', async () => {
    const tmpFile = path.join(__dirname, 'fixtures', 'basic-test.ts');
    const result = await Providers({
      ...migrateOpts,
      files: [tmpFile]
    });
    expect(result).toMatchSnapshot();
    parseCodeSnippets(result);
  });

  it('should migrate basic file with EslintProvider', async () => {
    const tmpFile = path.join(__dirname, 'fixtures', 'comma-test.ts');
    const result = await Providers({
      ...migrateOpts,
      files: [tmpFile]
    });
    expect(result).toMatchSnapshot();
    parseCodeSnippets(result);
  });

  it('should perform unsafe transformations', async () => {
    const result1 = await Providers({
      ...migrateOpts,
      unsafe: true,
      files: [
        path.join(__dirname, 'fixtures', 'unsafe-transformation-test.ts'),
        path.join(__dirname, 'fixtures', 'class-test.ts')
      ]
    });
    expect(result1).toMatchSnapshot();
    parseCodeSnippets(result1);

    const result2 = await Providers({
      ...migrateOpts,
      unsafe: true,
      files: [
        path.join(__dirname, 'fixtures', 'unsafe-transformation-test.ts'),
        path.join(__dirname, 'fixtures', 'class-test.ts')
      ]
    });
    expect(result2).toMatchSnapshot();
    parseCodeSnippets(result2);
  });

  it('should fail on non-existent files', () => {
    return expect(
      Providers({
        ...migrateOpts,
        files: [
          path.join(__dirname, 'fixtures', 'foo-test.ts'),
          path.join(__dirname, 'fixtures', 'class-test.ts')
        ]
      }).then(e => ({ ...e, path: '' }))
    ).rejects.toThrowError();
  });

  it('should run against directories', async () => {
    const result = await Providers({
      ...migrateOpts,
      files: [
        path.join(__dirname, 'fixtures', 'express-hello-world', 'foobar.js'),
        path.join(__dirname, 'fixtures', 'express-hello-world', 'index.js')
      ]
    });
    expect(result).toMatchSnapshot();
    parseCodeSnippets(result);
  });

  it("should not parse .gitignore'd files", async () => {
    const files = await handleInput({
      ...migrateOpts,
      files: [path.join(__dirname, './fixtures/express-hello-world')]
    });

    files.forEach(file => {
      chaiExpect(file).to.not.include('node_modules');
    });
  });

  // it.skip('should allow selection of manual tests', () => {});

  // it.skip('should not write files if one provider fails', async () => {});

  // it.skip('should fail on non-js extensions', async () => {});

  // it.skip('should fail on parsing coffeescript', async () => {});
});
