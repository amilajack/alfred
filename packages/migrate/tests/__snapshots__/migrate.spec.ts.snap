// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migrate should migrate basic file 1`] = `
Array [
  "import fs from 'fs';
import path from 'path';

const foo = 'asdfa';
console.log(foo);
",
]
`;

exports[`Migrate should migrate basic file with EslintProvider 1`] = `
Array [
  "const some = 'foo';
const baz = 'baz';
",
]
`;

exports[`Migrate should perform unsafe transformations 1`] = `
Array [
  "class MyClass {
  method() {}
}
",
  "import foo from 'bar';
import cow from 'boo';
",
]
`;

exports[`Migrate should perform unsafe transformations 2`] = `
Array [
  "class MyClass {
  method() {}
}
",
  "import foo from 'bar';
import cow from 'boo';
",
]
`;

exports[`Migrate should run against directories 1`] = `
Array [
  "/*!
 * express-hello-world <https://github.com/tunnckoCore/express-hello-world>
 *
 * Copyright (c) 2015 Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

/* jshint asi:true */
/* eslint import/no-unresolved: off, import/no-extraneous-dependencies: off */

import test from 'assertit';

import express from 'express';
import request from 'supertest';
import helloWorld from './index';

test('express-hello-world:', () => {
  test('should say \\"Hello World\\"', () => {
    return new Promise(done => {
      const app = express();
      app.use(helloWorld());

      request(app)
        .get('/')
        .expect(200, 'Hello World')
        .end(done);
    });
  });
  test('should yield next middleware', () => {
    return new Promise(done => {
      const app = express();
      app
        .use((req, res, next) => {
          res.set('X-First', 'foo');
          next();
        })
        .use(helloWorld())
        .use((req, res) => {
          test.equal(res.get('X-First'), 'foo');
        });

      request(app)
        .get('/')
        .expect(200, 'Hello World')
        .expect('X-First', 'foo')
        .end(err => {
          test.ifError(err);
          done();
        });
    });
  });
});
",
  "/*!
 * express-hello-world <https://github.com/tunnckoCore/express-hello-world>
 *
 * Copyright (c) 2015 Charlike Mike Reagent <@tunnckoCore> (http://www.tunnckocore.tk)
 * Released under the MIT license.
 */

export default function expressHelloWorld() {
  return (req, res, next) => {
    res.send('Hello World');
    next();
  };
}
",
]
`;
